rclcpp/src/rclcpp/executor.cpp

```diff
@@ -232,14 +232,20 @@ Executor::spin_some(std::chrono::nanoseconds max_duration)
     throw std::runtime_error("spin_some() called while already spinning");
   }
   RCLCPP_SCOPE_EXIT(this->spinning.store(false); );
-  // non-blocking call to pre-load all available work
-  wait_for_work(std::chrono::milliseconds::zero());
-  while (spinning.load() && max_duration_not_elapsed()) {
+  bool work_available = false;
+  while (rclcpp::ok(this->context_) && spinning.load() && max_duration_not_elapsed()) {
```

Comment:
The problem is that with the old `spin_some` implementation, no having that check wasn't much of a problem. It would take an unnoticeable time for `spin_some` to finish and react to `ctrl-c`.

Now that we're repeatedly waiting to check if there's more work, that time can be really noticeable.

I will split that change, so we can cherry-pick it separately.

Reply:
Ok makes sense.